File: python/armstrong-numbers/armstrong_numbers.py
"""
Module to check if a number is an Armstrong number.

An Armstrong number is a number that is equal to the sum of its digits 
raised to the power of the number of digits.
"""


def is_armstrong_number(number: int) -> bool:
    """
    Returns True if the number is an Armstrong number.
    """
    return _get_armstrong_value(number) == number


def _get_armstrong_value(number: int) -> int:
    """
    Calculates the sum of the digits raised to the power of the number of digits.
    This value is used to check if a number is an Armstrong number.
    """
    return sum(_get_armstrong_digits(number))


def _get_armstrong_digits(number: int) -> list[int]:
    """
    Raises each digit of the number to the power of the number of digits.
    """
    digits = _get_digits(number)
    return [digit ** len(digits) for digit in digits]


def _get_digits(number: int) -> list[int]:
    """
    Extracts and returns the digits of the number as a list of integers.
    """
    return [int(digit) for digit in str(number)]

File: go/armstrong-numbers/armstrong_numbers.go
// Package armstrong provides utilities to check if a number is an Armstrong number.
package armstrong

import (
	"math"
	"strconv"
)

// IsNumber checks if a given number is an Armstrong number.
// An Armstrong number is a number that is equal to the sum of its digits,
// each raised to the power of the number of digits.
func IsNumber(n int) bool {
	return armstrongSum(n) == n
}

// armstrongSum calculates the Armstrong sum of a number.
// The Armstrong sum is the sum of each digit of the number raised to the
// power of the total number of digits.
func armstrongSum(n int) int {
	digitsOfN := digits(n)
	digitsCount := len(digitsOfN)

	sum := 0
	for _, digit := range digitsOfN {
		sum += digitPow(digit, digitsCount)
	}

	return sum
}

// digits extracts the digits of a number and returns them as a slice of integers.
func digits(n int) []int8 {
	strNum := strconv.Itoa(n)
	digitsCount := len(strNum)
	digs := make([]int8, digitsCount)

	for i := 0; i < digitsCount; i++ {
		digs[i] = int8(strNum[i] - '0')
	}

	return digs
}

// digitPow raises a single digit to the specified power and returns the result.
func digitPow(digit int8, power int) int {
	return int(
		math.Pow(
			float64(digit),
			float64(power),
		),
	)
}
